### PR TITLE
S1 us6 formulaire

### PR DESCRIPTION
**Pull request by khadimb**

### CODE REVIEW

**Review by Vannou28 — COMMENTED** [review]
hello khadim, I'm sorry but you have too many magic values and the import of the font family in the css should be in the .css style.
sorry ;-(((

**Review by RaphaelBS-WCS — DISMISSED** [review]
I approve

File: public/assets/css/contact.css [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    }
+}
+
+@media (max-width: 768px) {
+    .subject {
+        width: 70%;
+    }
+}
+
+
```
Comment (LucyMerlier) [inline public/assets/css/contact.css]: There are still magic values here and there that could have been avoided

File: src/Controller/ContactController.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            if (empty($data['tel'])) {
+                $errors[] = 'Veuillez renseigner votre numéro de téléphone';
+            }
+            if (empty($data['subject'])) {
+                $errors[] = 'Veuillez renseigner le motif';
+            }
+            if (empty($data['message'])) {
+                $errors[] = 'Un message est obligatoire';
+            }
+        }
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:43]: incoherent error messages

**Review by RaphaelBS-WCS — DISMISSED** [review]
Approve

**Review by Vannou28 — APPROVED** [review]
it's good for me!!!

File: src/Controller/ContactController.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:44]: and what if there are no errors?

File: src/Controller/ContactController.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            if (empty($data['email'])) {
+                $errors[] = 'L\'email est obligatoire';
+            }
+            if (!filter_var($data['email'], FILTER_VALIDATE_EMAIL)) {
+                $errors[] = 'Mauvais format d\'email';
+            }
+            if (empty($data['tel'])) {
+                $errors[] = 'Veuillez renseigner votre numéro de téléphone';
+            }
+            if (empty($data['subject'])) {
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:37]: are there no other validations on the subject?

File: src/Controller/ContactController.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                $errors[] = 'Un nom complet est obligatoire';
+            } elseif ($data['name'] > MAX_LENGTH_NAME) {
+                $errors[] = 'Un nom complet doit faire moins de ' . MAX_LENGTH_NAME . ' caractères';
+            }
+            if (empty($data['email'])) {
+                $errors[] = 'L\'email est obligatoire';
+            }
+            if (!filter_var($data['email'], FILTER_VALIDATE_EMAIL)) {
+                $errors[] = 'Mauvais format d\'email';
+            }
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:33]: email length?

File: src/View/Contact/contact.html.twig [hunk 6]
```diff
@@ -0,0 +1,40 @@
+{% extends 'layout.html.twig' %}
+
+{% block stylesheet %}<link rel="stylesheet" href="/assets/css/contact.css">{% endblock %}
+{% block content %}
+
+    <form action="/contact/index" class="form" method= "POST" novalidate>
```
Comment (LucyMerlier) [inline src/View/Contact/contact.html.twig:6]: novalidate?

**Review by LucyMerlier — CHANGES_REQUESTED** [review]
see comments

File: src/View/Contact/contact.html.twig (hunk 6 above)
Comment (khadimb) [reply src/View/Contact/contact.html.twig:6]: ok

File: src/Controller/ContactController.php (hunk 2 above)
Comment (khadimb) [reply src/Controller/ContactController.php:43]: ok

File: public/assets/css/contact.css (hunk 1 above)
Comment (khadimb) [reply public/assets/css/contact.css]: ok

File: src/Controller/ContactController.php (hunk 3 above)
Comment (khadimb) [reply src/Controller/ContactController.php:44]: ok

File: src/Controller/ContactController.php (hunk 5 above)
Comment (khadimb) [reply src/Controller/ContactController.php:33]: Ok

File: src/Controller/ContactController.php (hunk 4 above)
Comment (khadimb) [reply src/Controller/ContactController.php:37]: ok

**Review by RaphaelBS-WCS — APPROVED** [review]
Approved

**Review by Mcbess-san — APPROVED** [review]

File: src/Controller/ContactController.php [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    private array $errors = [];
+
+    public function index()
+    {
+        $subjects = ['tel-session', 'cabinet-session', 'infos', 'other'];
+
+        if ($_SERVER['REQUEST_METHOD'] === 'POST') {
+            $data = array_map('trim', $_POST);
+            define('MAX_LENGTH_NAME', 255);
+            define('MAX_LENGTH_EMAIL', 320);
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:25]: you're working in OOP, these constants could be class constants :)

File: src/Controller/ContactController.php [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            $this->SetCheckSentenceEmpty($data['tel'], 'Veuillez renseigner votre numéro de téléphone');
+            $this->SetCheckSentenceEmpty($data['message'], 'Un message est obligatoire');
+            $this->SetCheckSentenceEmpty($data['email'], 'L\'email est obligatoire');
+
+            $this->setCheckWordSize($data['name'], MAX_LENGTH_NAME, 'Le nom complet doit faire moins de '
+                . MAX_LENGTH_NAME . ' caractères');
+            $this->setCheckWordSize($data['email'], MAX_LENGTH_EMAIL, 'Votre adresse mail doit faire moins de '
+                . MAX_LENGTH_EMAIL . ' caractères');
+
+            $this->setCheckWordPresenceInArray($data['subject'], $subjects, 'Le sujet saisie n\'est pas valable');
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:37]: typo "saisi"

File: src/Controller/ContactController.php [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            $this->setCheckFilterValidateEmail($data['email'], 'Mauvais format d\'email');
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:38]: incoherent error messages

File: src/Controller/ContactController.php [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+            if (empty($this->getErrors())) {
+                echo 'Votre message à bien été envoyé.';
+            }
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:42]: what is this wild echo doing in a controller?

File: src/Controller/ContactController.php [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    private function setErrors(string $errors)
+    {
+        $this->errors[] = $errors;
+    }
+
+    private function getErrors(): array
+    {
+        return $this->errors;
+    }
+}
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:86]: all those methods should be private if you want to let them in this class
+ warning, your class will soon have more than 10 methods => too much. Maybe refactor and put those methods in their own class?

File: src/Controller/ContactController.php [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+                echo 'Votre message à bien été envoyé.';
+            }
+        }
+
+        return $this->twig->render('Contact/contact.html.twig', ['errors' => $this->getErrors()]);
+    }
+
+    public function setCheckWordPresenceInArray(string $word, array $array, string $message)
+    {
+        var_dump($word);
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:50]: var_dump?

File: src/Controller/ContactController.php [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+            if (empty($this->getErrors())) {
+                echo 'Votre message à bien été envoyé.';
+            }
+        }
+
+        return $this->twig->render('Contact/contact.html.twig', ['errors' => $this->getErrors()]);
+    }
+
+    public function setCheckWordPresenceInArray(string $word, array $array, string $message)
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:48]: why "set" keywords at the start of the methods names? these are not setters

File: src/View/Contact/contact.html.twig (hunk 6 above)
Comment (LucyMerlier) [reply src/View/Contact/contact.html.twig:6]: no i meant, you probably don't want to keep this novalidate here

File: src/View/Contact/contact.html.twig [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+                    <option value="cabinet-session">Séance au cabinet</option>
+                    <option value="infos">Demande d'informations</option>
+                    <option value="other">Autres</option>
+                </select>
+            </div>
+            <div>
+                <label for="message">Votre message :</label>
+                <textarea name="message" id="message" rows="4" placeholder="Votre message" required></textarea>
+            </div>
+        <div class="buttoncontact">
```
Comment (LucyMerlier) [inline src/View/Contact/contact.html.twig:36]: use kebab-case for class names

**Review by LucyMerlier — CHANGES_REQUESTED** [review]
see comments

File: src/Controller/ContactController.php (hunk 12 above)
Comment (khadimb) [reply src/Controller/ContactController.php:50]: ok

File: src/View/Contact/contact.html.twig (hunk 14 above)
Comment (khadimb) [reply src/View/Contact/contact.html.twig:36]: ok

Comment (khadimb) [reply src/View/Contact/contact.html.twig:36]: ok

File: src/Controller/ContactController.php (hunk 12 above)
Comment (khadimb) [reply src/Controller/ContactController.php:50]: ok

File: src/Controller/ContactController.php (hunk 8 above)
Comment (khadimb) [reply src/Controller/ContactController.php:37]: ok

File: src/Controller/ContactController.php (hunk 10 above)
Comment (khadimb) [reply src/Controller/ContactController.php:42]: ok

File: src/Controller/ContactController.php (hunk 13 above)
Comment (khadimb) [reply src/Controller/ContactController.php:48]: ok

File: src/Controller/ContactController.php (hunk 11 above)
Comment (khadimb) [reply src/Controller/ContactController.php:86]: ok

File: src/Controller/ContactController.php (hunk 7 above)
Comment (khadimb) [reply src/Controller/ContactController.php:25]: ok

Comment (khadimb) [reply src/Controller/ContactController.php:25]: ok

File: src/Controller/ContactController.php (hunk 9 above)
Comment (khadimb) [reply src/Controller/ContactController.php]: ok

**Review by Vannou28 — DISMISSED** [review]
that is good !!!!

**Review by RaphaelBS-WCS — DISMISSED** [review]
Approve

File: src/Controller/ContactController.php [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+    public function index()
+    {
+        $subjects = ['tel-session', 'cabinet-session', 'infos', 'other'];
+
+        if ($_SERVER['REQUEST_METHOD'] === 'POST') {
+            $data = array_map('trim', $_POST);
+            $this->checkSentenceEmpty($data['name'], 'Un nom complet est obligatoire');
+            $this->checkSentenceEmpty($data['tel'], 'Un numéro de téléphone est obligatoire');
+            $this->checkSentenceEmpty($data['message'], 'Un message est obligatoire');
+            $this->checkSentenceEmpty($data['email'], 'L\'email est obligatoire');
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:29]: incoherent error messages

File: src/Controller/ContactController.php [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+
+            $this->checkWordSize($data['name'], self::MAX_LENGTH_NAME, 'Le nom complet doit faire moins de '
+                . self::MAX_LENGTH_NAME . ' caractères');
+            $this->checkWordSize($data['email'], self::MAX_LENGTH_EMAIL, 'Votre adresse mail doit faire moins de '
+                . self::MAX_LENGTH_EMAIL . ' caractères');
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:34]: incoherent error messages

File: src/Controller/ContactController.php [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+
+            $this->checkWordPresenceInArray($data['subject'], $subjects, 'Le sujet saisi n\'est pas valable');
+            $this->checkFilterValidateEmail($data['email'], 'Mauvais format d\'email');
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:37]: incoherent error messages

File: src/Controller/ContactController.php [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        if (empty($sentences)) {
+            $this->setErrors($messageError);
+        }
+    }
+
+    private function setErrors(string $errors)
+    {
+        $this->errors[] = $errors;
+    }
```
Comment (LucyMerlier) [inline src/Controller/ContactController.php:78]: no return typehint for all you form checkers?

File: src/View/Contact/contact.html.twig [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+            {% endfor %}
+            </ul>
+        {% endif %}
+        <div class="info">
+            <label for="name">Nom complet :</label>
+            <input class="name" type="text" name="name" placeholder="Nom complet" required>
+            <label for="email">Adresse mail :</label>
+            <input class="email" type="text" name="email" placeholder="exemple@exemple.com" required>
+            <label for="tel">Numéro de téléphone :</label>
+            <input class="tel" type="tel" name="tel" size="10" placeholder="00 00 00 00 00" required>
```
Comment (LucyMerlier) [inline src/View/Contact/contact.html.twig:21]: size 10 but your placeholder is 14 characters long?

File: src/View/Contact/contact.html.twig [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+                <option value="infos">Demande d'informations</option>
+                <option value="other">Autres</option>
+            </select>
+        </div>
+        <div>
+            <label for="message">Votre message :</label>
+            <textarea name="message" id="message" rows="4" placeholder="Votre message" required></textarea>
+        </div>
+        <div class="button-contact">
+            <button type="submit" href="">Envoyer</button>
```
Comment (LucyMerlier) [inline src/View/Contact/contact.html.twig:36]: why href?

**Review by LucyMerlier — CHANGES_REQUESTED** [review]

File: src/Controller/ContactController.php [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+     *
+     * @return string
+     * @throws \Twig\Error\LoaderError
+     * @throws \Twig\Error\RuntimeError
+     * @throws \Twig\Error\SyntaxError
+     */
+
+    private array $errors = [];
+    private const MAX_LENGTH_NAME = 255;
+    private const MAX_LENGTH_EMAIL = 320;
```
[the remark this answers is not in the document]
Comment (khadimb) [reply src/Controller/ContactController.php:18]: ok

File: src/View/Contact/contact.html.twig (hunk 19 above)
Comment (khadimb) [reply src/View/Contact/contact.html.twig:21]: ok

File: src/View/Contact/contact.html.twig (hunk 20 above)
Comment (khadimb) [reply src/View/Contact/contact.html.twig:36]: error ok

File: src/Controller/ContactController.php (hunk 15 above)
Comment (khadimb) [reply src/Controller/ContactController.php:29]: ok

File: src/Controller/ContactController.php (hunk 16 above)
Comment (khadimb) [reply src/Controller/ContactController.php:34]: ok

File: src/Controller/ContactController.php (hunk 17 above)
Comment (khadimb) [reply src/Controller/ContactController.php:37]: ok

File: src/Controller/ContactController.php [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+            $this->checkWordSize($data['name'], self::MAX_LENGTH_NAME, 'Le nom complet doit faire moins de '
+                . self::MAX_LENGTH_NAME . ' caractères');
+            $this->checkWordSize($data['email'], self::MAX_LENGTH_EMAIL, 'Votre adresse mail doit faire moins de '
+                . self::MAX_LENGTH_EMAIL . ' caractères');
+
+            $this->checkWordPresenceInArray($data['subject'], $subjects, 'Le sujet saisi n\'est pas valable');
+            $this->checkFilterValidateEmail($data['email'], 'Mauvais format d\'email');
+
+            if (empty($this->getErrors())) {
+                return 'Votre message à bien été envoyé.';
```
[the remark this answers is not in the document]
Comment (khadimb) [reply src/Controller/ContactController.php]: ok

File: src/Controller/ContactController.php (hunk 18 above)
Comment (khadimb) [reply src/Controller/ContactController.php:78]: ok

**Review by Mcbess-san — DISMISSED** [review]

File: src/View/Contact/success.html.twig [hunk 23]
```diff
@@ -0,0 +1,14 @@
...
+
+{% block stylesheet %}<link rel="stylesheet" href="/assets/css/success.css">{% endblock %}
+
+{% block content %}
+    <div class="box-container">
+        <div class="success-box">
+            <h2 class="success-message">Votre message à bien été envoyé</h2>
+            <form action="#" class="button-contact">
+                <button type="submit">Accueil</button>
+            </form>
```
Comment (LucyMerlier) [inline src/View/Contact/success.html.twig:11]: a form? shouldn't this be a link?

**Review by LucyMerlier — CHANGES_REQUESTED** [review]
see comment

**Review by RaphaelBS-WCS — DISMISSED** [review]
Approved!

File: src/View/Contact/success.html.twig [hunk 24]
```diff
@@ -0,0 +1,14 @@
+{% extends 'layout.html.twig' %}
+
+{% block stylesheet %}<link rel="stylesheet" href="/assets/css/success.css">{% endblock %}
+
+{% block content %}
+    <div class="box-container">
+        <div class="success-box">
+            <h2 class="success-message">Votre message à bien été envoyé</h2>
+            <a href="http://localhost:8000/" class="button-contact">
```
Comment (LucyMerlier) [inline src/View/Contact/success.html.twig:9]: NO! your route starts at the first "/" from your domain / host => if you want a link to go back to home page, just put "/" in href!

**Review by LucyMerlier — CHANGES_REQUESTED** [review]
see comments

File: src/View/Contact/success.html.twig [hunk 25]
```diff
@@ -0,0 +1,14 @@
...
+                <button type="submit">Accueil</button>
```
[the remark this answers is not in the document]
Comment (khadimb) [reply src/View/Contact/success.html.twig:10]: ok

File: src/View/Contact/success.html.twig (hunk 24 above)
Comment (khadimb) [reply src/View/Contact/success.html.twig:9]: ok

**Review by RaphaelBS-WCS — APPROVED** [review]
Approved

**Review by LucyMerlier — APPROVED** [review]

**Pull request merged**